### PR TITLE
Added admin audit log for admin config integration setup

### PR DESCRIPTION
**Pull request by SagarGi**
## Description
This PR:
  - Adds a way to log every activities regarding the admin integration configuration to log them in `audit.log` file in the Nextcloud data.
 
> Note: This PR does not implement the info warning in the UI. Since it is dependent to this bump PR: https://github.com/nextcloud/integration_openproject/pull/654.

## Related Issue or Workpackage

- Fixes https://community.openproject.org/projects/nextcloud-integration/work_packages/55110

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Updated `CHANGELOG.md` file

### CODE REVIEW

**Review by individual-it — COMMENTED** [review]
is this something we want to have in 2.6 or only in 2.7?

File: lib/Service/OpenProjectAPIService.php [hunk 1]
```diff
@@ -1070,6 +1078,33 @@ class_exists('\OCA\GroupFolders\Folder\FolderManager') &&
 		);
 	}
 
+	/**
+	 * @param $auditLogMessage
+	 * @return void
+	 */
+	public function logToAuditFile($auditLogMessage): void {
+		if($this->isAdminAuditConfigSetCorrectly()) {
```
Comment (individual-it) [inline lib/Service/OpenProjectAPIService.php:1086]: if the audit log is not setup correctly, would it make sense to log to the normal log file?

**Comment by SagarGi** [comment]
> is this something we want to have in 2.6 or only in 2.7?

we will be having a patch release.

File: tests/lib/Service/OpenProjectAPIServiceTest.php [hunk 2]
```diff
@@ -3607,4 +3619,97 @@ public function testGetWorkPackageInfoForNoUserAccessToken(): void {
...
+	 * @param int $logLevel
+	 * @param string $pathToAuditLog
+	 * @param array<mixed> $logCondition
+	 * @param string $isAdminAuditAppInstalled
+	 * @param bool $expectedResult
+	 *
+	 * @return void
+	 */
+
+	public function testIsAdminAuditConfigSetCorrectly(int $logLevel, string $pathToAuditLog, array $logCondition, bool $isAdminAuditAppInstalled, bool $expectedResult): void {
```
Comment (individual-it) [inline tests/lib/Service/OpenProjectAPIServiceTest.php:3687]: too long line

File: lib/Controller/OpenProjectAPIController.php [hunk 3]
```diff
@@ -547,6 +547,9 @@ public function isValidOpenProjectInstance(string $url): DataResponse {
 			);
 			return new DataResponse(['result' => 'invalid']);
 		}
+		$oldOpenProjectOauthUrl = $this->config->getAppValue(
```
Comment (individual-it) [inline lib/Controller/OpenProjectAPIController.php:550]: seems to be not used, why did the CI not catch it?

File: lib/Controller/ConfigController.php [hunk 4]
```diff
@@ -337,6 +352,9 @@ private function setIntegrationConfig(array $values): array {
 		// so setting `fresh_project_folder_setup` as true
 		if ($runningFullReset) {
 			$this->config->setAppValue(Application::APP_ID, 'fresh_project_folder_setup', "1");
+			$this->openprojectAPIService->logToAuditFile(
+				"OpenProject Integration admin configuration has been reset in application " . Application::APP_ID, true
```
Comment (individual-it) [inline lib/Controller/ConfigController.php:356]: there is no second parameter in `logToAuditFile`, why is CI not catching that?

File: CHANGELOG.md [hunk 5]
```diff
@@ -7,6 +7,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 ## [Unreleased]
 ### Changed
 - Add application's support for Nextcloud 30
+- Audit logging in `/audit.log` for admin configuration.
```
Comment (individual-it) [inline CHANGELOG.md:10]:
```suggestion
- log admin configuration to  audit log (`/audit.log`) 
```

File: lib/Service/OpenProjectAPIService.php [hunk 6]
```diff
@@ -49,9 +50,11 @@
 use OCP\IL10N;
 use OCP\IURLGenerator;
 use OCP\IUserManager;
+use OCP\Log\ILogFactory;
 use OCP\PreConditionNotMetException;
 use OCP\Security\ISecureRandom;
 use OCP\Server;
+use phpDocumentor\Reflection\Types\This;
```
Comment (individual-it) [inline lib/Service/OpenProjectAPIService.php:57]: what is that for?

**Review by individual-it — CHANGES_REQUESTED** [review]

File: lib/Controller/OpenProjectAPIController.php (hunk 3 above)
Comment (nabim777) [reply lib/Controller/OpenProjectAPIController.php:550]: There is a WP create that will address this CI [issue](https://community.openproject.org/projects/nextcloud-integration/work_packages/55940/activity?query_props=%7B%22c%22%3A%5B%22id%22%2C%22subject%22%2C%22type%22%2C%22status%22%2C%22assignee%22%5D%2C%22hi%22%3Atrue%2C%22g%22%3A%22%22%2C%22is%22%3Atrue%2C%22tv%22%3Afalse%2C%22hla%22%3A%5B%22status%22%2C%22priority%22%2C%22dueDate%22%5D%2C%22t%22%3A%22id%3Aasc%22%2C%22f%22%3A%5B%7B%22n%22%3A%22status%22%2C%22o%22%3A%22o%22%2C%22v%22%3A%5B%5D%7D%5D%2C%22ts%22%3A%22PT0S%22%2C%22pp%22%3A50%2C%22pa%22%3A2%7D).

File: lib/Service/OpenProjectAPIService.php (hunk 6 above)
Comment (SagarGi) [reply lib/Service/OpenProjectAPIService.php]: Auto imported. Deleting it as it is not required.

File: lib/Controller/ConfigController.php (hunk 4 above)
Comment (SagarGi) [reply lib/Controller/ConfigController.php]: will be solved by this WP [issue](https://community.openproject.org/projects/nextcloud-integration/work_packages/55940/activity?query_props=%7B%22c%22%3A%5B%22id%22%2C%22subject%22%2C%22type%22%2C%22status%22%2C%22assignee%22%5D%2C%22hi%22%3Atrue%2C%22g%22%3A%22%22%2C%22is%22%3Atrue%2C%22tv%22%3Afalse%2C%22hla%22%3A%5B%22status%22%2C%22priority%22%2C%22dueDate%22%5D%2C%22t%22%3A%22id%3Aasc%22%2C%22f%22%3A%5B%7B%22n%22%3A%22status%22%2C%22o%22%3A%22o%22%2C%22v%22%3A%5B%5D%7D%5D%2C%22ts%22%3A%22PT0S%22%2C%22pp%22%3A50%2C%22pa%22%3A2%7D).

**Comment by SagarGi** [comment]
@nabim777 can you manually test it once.

File: lib/Controller/ConfigController.php [hunk 7]
```diff
@@ -528,7 +555,11 @@ private function storeUserInfo(): array {
 	 * @return DataResponse
 	 */
 	public function autoOauthCreation(): DataResponse {
-		return new DataResponse($this->recreateOauthClientInformation());
+		$result = $this->recreateOauthClientInformation();
+		$this->openprojectAPIService->logToAuditFile(
+			"Nextcloud OAuth credential has been set on application " . Application::APP_ID
```
Comment (individual-it) [inline lib/Controller/ConfigController.php:560]:
```suggestion
			"Nextcloud OAuth credential has been set in application " . Application::APP_ID
```

File: lib/Controller/ConfigController.php [hunk 8]
```diff
@@ -366,6 +384,15 @@ private function setIntegrationConfig(array $values): array {
 	public function setAdminConfig(array $values): DataResponse {
 		try {
 			$result = $this->setIntegrationConfig($values);
+			if(key_exists('openproject_client_id', $values) &&
+				key_exists('openproject_client_secret', $values) &&
+				$values['openproject_client_id'] &&
+				$values['openproject_client_secret']
+			) {
+				$this->openprojectAPIService->logToAuditFile(
```
Comment (individual-it) [inline lib/Controller/ConfigController.php:392]: what about logging what keys have been set?

File: lib/Controller/OpenProjectAPIController.php [hunk 9]
```diff
@@ -585,6 +585,10 @@ public function isValidOpenProjectInstance(string $url): DataResponse {
 				$decodedBody['_type'] === 'Root' &&
 				$decodedBody['instanceName'] !== ''
 			) {
+				// sending admin audit log if admin has changed or added the openproject host url
+				$this->openprojectAPIService->logToAuditFile(
+					"OpenProject host url has been set to '$url' in application " . Application::APP_ID
```
Comment (individual-it) [inline lib/Controller/OpenProjectAPIController.php:590]: in this function we don't set the URL, but only check if its correct, we should log the things only when they really happen

**Review by individual-it — CHANGES_REQUESTED** [review]

File: lib/Controller/ConfigController.php (hunk 8 above)
Comment (SagarGi) [reply lib/Controller/ConfigController.php:392]: would be nice to have it as well.

File: lib/Controller/OpenProjectAPIController.php (hunk 9 above)
Comment (SagarGi) [reply lib/Controller/OpenProjectAPIController.php]: True changing it right away!

**Review by individual-it — APPROVED** [review]

**Review by nabim777 — APPROVED** [review]
Manually checked in server stable-26, 27, and 29 It works fine there

**Pull request merged**